nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -0,0 +1,534 @@
+
+  int64_t i;
+  ml_logd("<called>");
+
+  num_tensors = prop->input_meta.num_tensors;
+  ml_logd("num_tensors: %" PRId64 "", num_tensors);
+
+  for (i = 0; i < num_tensors; i++) {
+    tensors_inputsize[i] = gst_tensor_info_get_size(&prop->input_meta.info[i]);
+    ml_logd("tensors_inputsize[%" PRId64 "]:%" PRId64 "", i,
+            tensors_inputsize[i]);
+  }
+  // for mnist test
+#if 0
+  tensors_inputsize[1] = 40;
+  // tensors_inputsize[0] = 3686400; // 3:640:480:1 float32, 3x640x480x1x4
+  // tensors_inputsize[1] = 40;      // 1:1:10:1 uint8, 1x1x10x1x4
+  ml_logd("for Test: tensors_inputsize[1]:%d", tensors_inputsize[1]);
+#endif
+  num_inputs = prop->num_inputs;
+  num_labels = prop->num_labels;
+  num_train_samples = prop->num_train_samples;
+  num_valid_samples = prop->num_valid_samples;
+  model_save_path = prop->model_save_path;
+  num_epochs = prop->num_epochs;
+  train_complete_cond = prop->train_complete_cond;
+  is_train_complete = FALSE;
+  total_num_samples = (num_train_samples + num_valid_samples) * num_epochs;
+
+  ml_logd("num_inputs: %" PRId64 "", num_inputs);
+  ml_logd("num_labels: %" PRId64 "", num_labels);
+  ml_logd("num_train_samples: %" PRId64 "", num_train_samples);
+  ml_logd("num_valid_samples: %" PRId64 "", num_valid_samples);
+  ml_logd("num_epochs: %" PRId64 "", num_epochs);
+  ml_logd("Total number of data to be received: %" PRId64 "",
+          total_num_samples);
+  ml_logd("model_config: %s", model_config.c_str());
+  ml_logd("model_save_path: %s", model_save_path.c_str());
+  ml_logd("<leave>");
+}
+
+static int nntrainer_model_destructor(const GstTensorTrainerFramework *fw,
+                                      const GstTensorTrainerProperties *prop,
+                                      void **private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    static_cast<NNTrainer::NNTrainerTrain *>(*private_data);
+  UNUSED(fw);
+  ml_logd("<called>");
+
+  if (!nntrainer)
+    return -1;
+
+  delete nntrainer;
+  *private_data = NULL;
+  ml_logd("<leave>");
+
+  return 0;
+}
+
+static int
+nntrainer_model_start_training(const GstTensorTrainerFramework *fw,
+                               const GstTensorTrainerProperties *prop,
+                               void *private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    reinterpret_cast<NNTrainer::NNTrainerTrain *>(private_data);
+  UNUSED(fw);
+
+  ml_logd("<called>");
+  if (!nntrainer) {
+    ml_loge("Failed get nntrainer");
+  }
+  try {
+    std::thread train_thread(nntrainer_thread_func, nntrainer);
+    train_thread.detach();
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return -1;
+  }
+  ml_logd("<leave>");
+  return 0;
+}
+
+void NNTrainer::NNTrainerTrain::trainModel() {
+  pid_t pid = getpid();
+  pid_t tid = syscall(SYS_gettid);
+
+  ml_logd("<called>");
+  ml_logd("pid[%d], tid[%d]", pid, tid);
+
+  try {
+    model->setProperty({"epochs=" + std::to_string(num_epochs)});
```

Comment:
model property is not just epochs. Others can be added like batch_size.. and more.  Is there any reason adding the epoch only?

Reply:
As if talking to each other offline
When the data entering nntrainer exceeds the batch size,
it is necessary to check on the nntrainer side how it is processed.
The subplugin will handle exceptions if necessary.